source/secure_tunneling.c

```diff
@@ -473,198 +396,1395 @@ bool secure_tunneling_send_data_call(struct aws_websocket *websocket, struct aws
+ * instantaneously before final destruction.
+ */
+static uint64_t s_compute_next_service_time_secure_tunnel_stopped(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    /* have we been told to connect or terminate? */
+    if (secure_tunnel->desired_state != AWS_STS_STOPPED) {
+        return now;
+    }
+
+    return 0;
+}
+
+static uint64_t s_compute_next_service_time_secure_tunnel_connecting(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    (void)secure_tunnel;
+    (void)now;
+
+    return 0;
+}
+
+static uint64_t s_compute_next_service_time_secure_tunnel_connected(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    /* TODO check against ping timeout once pong is implemented by secure tunnel service */
+    uint64_t next_service_time = secure_tunnel->next_ping_time;
+
+    if (secure_tunnel->desired_state != AWS_STS_CONNECTED) {
+        next_service_time = now;
+    }
+
+    uint64_t operation_processing_time = s_aws_secure_tunnel_compute_operational_state_service_time(secure_tunnel, now);
+
+    next_service_time = s_min_non_0_64(operation_processing_time, next_service_time);
+
+    return next_service_time;
+}
+
+static uint64_t s_compute_next_service_time_secure_tunnel_clean_disconnect(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    return s_aws_secure_tunnel_compute_operational_state_service_time(secure_tunnel, now);
+}
+
+static uint64_t s_compute_next_service_time_secure_tunnel_websocket_shutdown(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    (void)secure_tunnel;
+    (void)now;
+
+    return 0;
+}
+
+static uint64_t s_compute_next_service_time_secure_tunnel_pending_reconnect(
+    struct aws_secure_tunnel *secure_tunnel,
+    uint64_t now) {
+    if (secure_tunnel->desired_state != AWS_STS_CONNECTED) {
+        return now;
+    }
+
+    return secure_tunnel->next_reconnect_time_ns;
+}
+
+static uint64_t s_compute_next_service_time_by_current_state(struct aws_secure_tunnel *secure_tunnel, uint64_t now) {
+
+    switch (secure_tunnel->current_state) {
+        case AWS_STS_STOPPED:
+            return s_compute_next_service_time_secure_tunnel_stopped(secure_tunnel, now);
+        case AWS_STS_CONNECTING:
+            return s_compute_next_service_time_secure_tunnel_connecting(secure_tunnel, now);
+        case AWS_STS_CONNECTED:
+            return s_compute_next_service_time_secure_tunnel_connected(secure_tunnel, now);
+        case AWS_STS_CLEAN_DISCONNECT:
+            return s_compute_next_service_time_secure_tunnel_clean_disconnect(secure_tunnel, now);
+        case AWS_STS_WEBSOCKET_SHUTDOWN:
+            return s_compute_next_service_time_secure_tunnel_websocket_shutdown(secure_tunnel, now);
+        case AWS_STS_PENDING_RECONNECT:
+            return s_compute_next_service_time_secure_tunnel_pending_reconnect(secure_tunnel, now);
+        case AWS_STS_TERMINATED:
+            return 0;
```

Comment:
Minor, debatable: I would either add a helper function for terminated, or have `s_compute_next_service_time_secure_tunnel_connecting` and `s_compute_next_service_time_secure_tunnel_websocket_shutdown` just return `0` as well, since (to the best of my knowledge) they are just placeholders.

Reply:
Yea, I thought about that too. This section is modeled after how we set up mqtt5 and I think the idea is that for terminated, it's forever and definitely going to be 0 but in those other cases, something could happen someday through some kind of addition of functionality?

Reply:
Hmm, maybe? I suppose if we do that in MQTT5, then it is probably fine and we can just extend/change it as needed - it just stood out to me when I was looking over it.